### PR TITLE
Kotlin: Sealed class not exaustive when checked in statement

### PR DESCRIPTION
**Pull request by mangatmodi**

### CODE REVIEW

File: src/types/sealedClass/sealedClass.kts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package types.sealedClass
```
Comment (mangatmodi) [inline src/types/sealedClass/sealedClass.kts:1]: Sealed class looks like has to have nested sub-classes when writing kotlin script (kts!)

**Comment by mangatmodi** [comment]
I found this case counter intuitive to the defination of sealed class. Looks like a special case, hence deserved a puzzler.

File: src/types/sealedClass/Rationale.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Correct answer: **a) Yes**
+
+* Exhaustive check on sealed classes does not happen when it is checked in a statement. Convert the check into expression as:- 
+```
+val out = when (h) {
+    is Dog -> println("It is a Dog")
+}
+```
+* More info at [Kotlin Sealed Classes](http://kotlinlang.org/docs/reference/sealed-classes.html)
```
Comment (TWiStErRob) [inline src/types/sealedClass/Rationale.md:9]: Long blog post about this "feature": https://proandroiddev.com/til-when-is-when-exhaustive-31d69f630a8b

File: src/types/sealedClass/Rationale.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Correct answer: **a) Yes**
+
+* Exhaustive check on sealed classes does not happen when it is checked in a statement. Convert the check into expression as:- 
+```
+val out = when (h) {
+    is Dog -> println("It is a Dog")
+}
+```
```
Comment (TWiStErRob) [inline src/types/sealedClass/Rationale.md:8]: note: statements can be made exhaustive in another way too: call a function or property on them
```
when (h) {
    is Animal.Dog -> println("It is a Dog")
}.javaClass
```
although this looks tricky and unrelated, it's better to use this really nice addition to Kotlin
```
val <T> T.exhaustive: T get() = this
```
from the blog post below (or above depending on which GitHub view you're using 😕)

**Review by TWiStErRob — COMMENTED** [review]
I think it deserves more info, this bit us as well. Added info as comments, feel free to copy them in.

File: src/types/sealedClass/Rationale.md (hunk 3 above)
Comment (mangatmodi) [reply src/types/sealedClass/Rationale.md:8]: Agreed `val <T> T.exhaustive: T get() = this` is much nicer as `javaclass` doesn't tell much about the reason for use. However I really don't like adding extension function to every type. We can also convert statements to expressions by passing them as lambda. e.g. ->
```
fun <T> exhaustive(func: () -> T) {
    func()
}

fun test_sealed_class_correct() {
    val h = Human() as Animal
    exhaustive {
        when (h) {
            is Animal.Dog -> null
        //Gives error!
        }
    }
}

```

Comment (TWiStErRob) [reply src/types/sealedClass/Rationale.md:8]: Nice idea as well, thanks for sharing; interesting that even though the return value is apparently (apart from being last expression in lambda) not used, it's still exhaustive. I agree about polluting the namespace.

Note there's also this: https://youtrack.jetbrains.com/issue/KT-12380

Just a bonkers idea for fun, cause nesting is ugly:
```kotlin
@Suppress("ClassName", "NOTHING_TO_INLINE")
object an {
    inline infix fun <T> exhaustive(whenValue: T) = whenValue
}

fun test_sealed_class_correct() {
    val h = Human() as Animal
    an exhaustive when (h) {
        is Animal.Dog -> null
        //Gives error!
    }
}
```

**Pull request merged**